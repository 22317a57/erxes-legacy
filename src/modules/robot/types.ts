export type IFeature = {
  name: string;
  text: string;
};
