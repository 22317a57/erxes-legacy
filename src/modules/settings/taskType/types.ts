export interface ITaskType {
  _id: string;
  name: string;
  icon: string;
}

// query types

export type TaskTypeQueryResponse = {
  taskTypes: ITaskType[];
  loading: boolean;
  refetch: () => void;
};

export type MutationVariables = {
  _id?: string;
  name?: string;
  icon?: string;
};

// mutation types

export type AddMutationResponse = {
  addMutation: (mutation: { variables: MutationVariables }) => Promise<any>;
};

export type EditMutationResponse = {
  editMutation: (mutation: { variables: MutationVariables }) => Promise<any>;
};

export type RemoveMutationResponse = {
  removeMutation: (mutation: { variables: { _id: string } }) => Promise<any>;
};
