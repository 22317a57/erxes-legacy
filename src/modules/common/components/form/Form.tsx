import * as React from 'react';
import { Form } from 'react-final-form';

type Props = {
  onSubmit: (e: any) => void;
  children: React.ReactNode;
  validate?: (values: any) => any;
};

function FormComponent({ onSubmit, validate, children }: Props) {
  const content = () => children;

  return (
    <Form
      onSubmit={onSubmit}
      validate={validate && validate}
      render={content}
    />
  );
}

export default FormComponent;
