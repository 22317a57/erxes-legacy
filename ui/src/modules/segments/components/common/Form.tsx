import Button from 'modules/common/components/Button';
import FormControl from 'modules/common/components/form/Control';
import CommonForm from 'modules/common/components/form/Form';
import FormGroup from 'modules/common/components/form/Group';
import ControlLabel from 'modules/common/components/form/Label';
import { IButtonMutateProps, IFormProps } from 'modules/common/types';
import { __, generateRandomColorCode } from 'modules/common/utils';
import { FlexContent, FlexItem } from 'modules/layout/styles';
import {
  IConditionFilter,
  IEvent,
  ISegment,
  ISegmentCondition,
  ISegmentWithConditionDoc
} from 'modules/segments/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { ConditionWrapper, SegmentTitle, SegmentWrapper } from '../styles';
import AddConditionButton from './AddConditionButton';
import EventCondition from './EventCondition';
import PropertyCondition from './PropertyCondition';

type Props = {
  contentType?: string;
  fields: any[];
  events: IEvent[];
  renderButton: (props: IButtonMutateProps) => JSX.Element;
  edit?: (params: { _id: string; doc: ISegmentWithConditionDoc }) => void;
  segment?: ISegment;
  headSegments: ISegment[];
  count: (segment: ISegment) => void;
  isForm?: boolean;
  afterSave?: () => void;
};

type State = {
  name: string;
  description: string;
  subOf: string;
  color: string;
  conditions: ISegmentCondition[];
};

class Form extends React.Component<Props, State> {
  constructor(props) {
    super(props);

    const segment: ISegment = props.segment || {
      name: '',
      description: '',
      subOf: '',
      color: generateRandomColorCode(),
      conditions: [],
    };

    segment.conditions = segment.conditions.map(
      (cond: ISegmentCondition) => ({
        key: Math.random().toString(),
        ...cond
      })
    );

    this.state = segment;
  }

  addCondition = (condition: ISegmentCondition) => {
    this.setState({
      conditions: [...this.state.conditions, condition]
    });
  };

  changeEventCondition = (args: { key: string, name: string, attributeFilters: IConditionFilter[] }) => {
    const condition = {
      type: 'event',
      key: args.key,
      eventName: args.name,
      eventAttributeFilters: (args.attributeFilters || []).map(filter => {
        const { key, ...rest } = filter;

        return rest;
      })
    }

    this.setState({
      conditions: this.state.conditions.map(c =>
        c.key === condition.key ? condition : c
      )
    });
  };

  changePropertyCondition = (args: { key: string, name: string, operator: string, value: string }) => {
    const condition = {
      type: 'property',
      key: args.key,
      propertyName: args.name,
      propertyOperator: args.operator,
      propertyValue: args.value,
    }

    this.setState({
      conditions: this.state.conditions.map(c =>
        c.key === condition.key ? condition : c
      )
    });
  };

  removeCondition = (key: string) => {
    const conditions = this.state.conditions.filter(c => c.key !== key);

    this.setState({ conditions });
  };

  handleChange = <T extends keyof State>(name: T, value: State[T]) => {
    this.setState({ [name]: value } as Pick<State, keyof State>);
  };

  generateDoc = (values: {
    _id?: string;
    name: string;
    subOf: string;
    color: string;
  }) => {
    const { segment, contentType } = this.props;
    const { conditions } = this.state;
    const finalValues = values;

    const updatedConditions: ISegmentCondition[] = [];

    if (segment) {
      finalValues._id = segment._id;
    }

    conditions.forEach((cond: ISegmentCondition) => {
      const { key, ...rest } = cond;
      updatedConditions.push(rest);
    });

    return {
      ...finalValues,
      contentType,
      conditions: updatedConditions
    };
  };

  renderParent() {
    const { contentType } = this.props;
    const { subOf } = this.state;

    if (!subOf) {
      return null;
    }

    return (
      <React.Fragment>
        <Link
          to={`/segments/edit/${contentType}/${subOf}`}
          target="_blank"
        >
          <Button icon="eye" ignoreTrans={true}>
            {__('Parent segment conditions')}
          </Button>
        </Link>
        <hr />
      </React.Fragment>
    );
  }

  renderCondition(condition: ISegmentCondition) {
    const { fields, events } = this.props;

    if (condition.type === 'property') {
      return (
        <PropertyCondition
          fields={fields}
          key={condition.key}
          conditionKey={condition.key || ''}
          name={condition.propertyName || ''}
          operator={condition.propertyOperator || ''}
          value={condition.propertyValue || ''}
          onChange={this.changePropertyCondition}
          onRemove={this.removeCondition}
        />
      )
    }

    return (
      <EventCondition
        events={events}
        key={condition.key}
        conditionKey={condition.key || ''}
        name={condition.eventName || ''}
        attributeFilters={condition.eventAttributeFilters || []}
        onChange={this.changeEventCondition}
        onRemove={this.removeCondition}
      />
    )
  }

  renderConditions() {
    const { conditions } = this.state;

    return (
      <React.Fragment>
        <ConditionWrapper>
          {this.renderParent()}
          {conditions.map(condition => this.renderCondition(condition))}
          </ConditionWrapper>
        <AddConditionButton addCondition={this.addCondition} />
      </React.Fragment>
    );
  }

  renderSubOf(formProps: IFormProps) {
    const onChange = (e: React.FormEvent) =>
      this.handleChange('subOf', (e.currentTarget as HTMLInputElement).value);

    return (
      <FormGroup>
        <ControlLabel>Sub segment of</ControlLabel>
        <FormControl
          {...formProps}
          name="subOf"
          componentClass="select"
          value={this.state.subOf || ''}
          onChange={onChange}
        >
          <option value="">[not selected]</option>
          {this.props.headSegments.map(segment => (
            <option value={segment._id} key={segment._id}>
              {segment.name}
            </option>
          ))}
        </FormControl>
      </FormGroup>
    );
  }

  renderForm = (formProps: IFormProps) => {
    const {
      isForm,
      segment,
      contentType,
      renderButton,
      afterSave
    } = this.props;

    const { values, isSubmitted } = formProps;
    const { name, description, color } = this.state;

    const nameOnChange = (e: React.FormEvent) =>
      this.handleChange('name', (e.currentTarget as HTMLInputElement).value);

    const descOnChange = (e: React.FormEvent) =>
      this.handleChange(
        'description',
        (e.currentTarget as HTMLInputElement).value
      );

    const colorOnChange = (e: React.FormEvent) =>
      this.handleChange('color', (e.currentTarget as HTMLInputElement).value);

    return (
      <>
        <FormGroup>
          <ControlLabel required={true}>Name</ControlLabel>
          <FormControl
            {...formProps}
            name="name"
            value={name}
            onChange={nameOnChange}
            required={true}
            autoFocus={true}
          />
        </FormGroup>
        <FormGroup>
          <ControlLabel>Description</ControlLabel>
          <FormControl
            {...formProps}
            name="description"
            value={description}
            onChange={descOnChange}
          />
        </FormGroup>
        {this.renderSubOf({ ...formProps })}
        <FormGroup>
          <ControlLabel>Color</ControlLabel>
          <FormControl
            {...formProps}
            name="color"
            type="color"
            value={color}
            onChange={colorOnChange}
          />
        </FormGroup>
        <Button.Group>
          {isForm && (
            <Link to={`/segments/${contentType}`}>
              <Button size="small" btnStyle="simple" icon="cancel-1">
                Cancel
              </Button>
            </Link>
          )}

          {renderButton({
            name: 'segment',
            values: this.generateDoc(values),
            callback: afterSave,
            isSubmitted,
            object: segment
          })}
        </Button.Group>
      </>
    );
  };

  render() {
    return (
      <SegmentWrapper>
        <SegmentTitle>{__('Filters')}</SegmentTitle>
        {this.renderConditions()}
        <hr />
        <FlexContent>
          <FlexItem count={4}>
            <CommonForm renderContent={this.renderForm} />
          </FlexItem>
          <FlexItem count={2} />
        </FlexContent>
      </SegmentWrapper>
    );
  }
}

export default Form;
