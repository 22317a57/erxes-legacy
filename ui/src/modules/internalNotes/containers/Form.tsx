import { useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import React, { useState } from 'react';
import Form from '../components/Form';
import { mutations } from '../graphql';
import {
  InternalNotesAddMutationResponse,
  InternalNotesAddMutationVariables
} from '../types';

type Props = {
  contentType: string;
  contentTypeId: string;
};

type State = {
  isLoading: boolean;
};

function FormContainer(props: Props, state: State) {
  const { contentType, contentTypeId } = props;
  const [isLoading, setLoading] = useState(false);

  const [internalNotesAdd] = useMutation<InternalNotesAddMutationResponse, InternalNotesAddMutationVariables>(gql(mutations.internalNotesAdd), {
    refetchQueries: ['activityLogs']
  });

  // create internalNote
  const create = (variables, callback: () => void) => {
    setLoading(true);

    internalNotesAdd({
      variables: {
        contentType,
        contentTypeId,
        ...variables
      }
    }).then(() => {
      callback();

      setLoading(false);
    });
  };

  return <Form save={create} isActionLoading={isLoading} />;
}
export default FormContainer;
