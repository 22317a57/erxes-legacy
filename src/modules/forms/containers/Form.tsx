import gql from 'graphql-tag';
import { Alert, withProps } from 'modules/common/utils';
import {
  AddIntegrationMutationResponse,
  AddIntegrationMutationVariables
} from 'modules/settings/integrations/types';
import {
  AddFieldsMutationResponse,
  AddFieldsMutationVariables
} from 'modules/settings/properties/types';
import React from 'react';
import { compose, graphql } from 'react-apollo';
import { withRouter } from 'react-router';
import { IRouterProps } from '../../common/types';
import Form from '../components/Form';
import { mutations } from '../graphql';
import {
  AddFormMutationResponse,
  AddFormMutationVariables,
  IFormPreviewContent
} from '../types';

type Props = {
  previewContent: (props: IFormPreviewContent) => void;
};

type FinalProps = {} & Props &
  IRouterProps &
  AddIntegrationMutationResponse &
  AddFieldsMutationResponse &
  AddFormMutationResponse;

class FormContainer extends React.Component<
  FinalProps,
  { isLoading: boolean }
> {
  constructor(props: FinalProps) {
    super(props);

    this.state = { isLoading: false };
  }

  render() {
    const {
      addIntegrationMutation,
      addFormMutation,
      addFieldsMutation,
      history
    } = this.props;

    const save = doc => {
      let leadId;

      const { form, brandId, name, languageCode, leadData, fields } = doc;

      this.setState({ isLoading: true });

      addFormMutation({
        variables: form
      })
        .then(({ data }) => {
          // tslint:disable-next-line:no-console
          console.log(data);
          leadId = data.formsAdd._id;

          return addIntegrationMutation({
            variables: { leadData, brandId, name, languageCode, leadId }
          });
        })

        .then(() => {
          const promises: any[] = [];

          for (const field of fields) {
            promises.push(
              addFieldsMutation({
                variables: {
                  contentType: 'form',
                  contentTypeId: leadId,
                  ...field
                }
              })
            );
          }

          return Promise.all(promises);
        })

        .then(() => {
          Alert.success('You successfully added a lead');
          history.push('/leads');

          this.setState({ isLoading: false });
        })

        .catch(error => {
          Alert.error(error.message);

          this.setState({ isLoading: false });
        });
    };

    const updatedProps = {
      ...this.props,
      fields: []
    };

    return <Form {...updatedProps} />;
  }
}

export default withProps<Props>(
  compose(
    graphql<
      Props,
      AddIntegrationMutationResponse,
      AddIntegrationMutationVariables
    >(gql(mutations.integrationsCreateLeadintegration), {
      name: 'addIntegrationMutation',
      options: {
        refetchQueries: ['formIntegrations', 'formIntegrationCounts']
      }
    }),
    graphql<Props, AddFormMutationResponse, AddFormMutationVariables>(
      gql(mutations.addForm),
      {
        name: 'addFormMutation'
      }
    ),
    graphql<Props, AddFieldsMutationResponse, AddFieldsMutationVariables>(
      gql(mutations.fieldsAdd),
      {
        name: 'addFieldsMutation'
      }
    )
  )(withRouter<FinalProps>(FormContainer))
);
