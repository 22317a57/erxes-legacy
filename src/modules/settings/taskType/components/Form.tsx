import Button from 'modules/common/components/Button';
import FormControl from 'modules/common/components/form/Control';
import CommonForm from 'modules/common/components/form/Form';
import FormGroup from 'modules/common/components/form/Group';
import ControlLabel from 'modules/common/components/form/Label';
import Icon from 'modules/common/components/Icon';
import { IButtonMutateProps, IFormProps } from 'modules/common/types';
import React from 'react';
import { Modal } from 'react-bootstrap';
import Select from 'react-select-plus';
import { TASK_ICONS } from '../constants';
import { ITaskType } from '../types';

type Props = {
  taskType?: ITaskType;
  renderButton: (props: IButtonMutateProps) => JSX.Element;
  closeModal: () => void;
};

type State = {
  selectedIcon: string;
};

class Form extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      selectedIcon: this.getSelectedIcon()
    };
  }

  onChangeIcon = obj => {
    this.setState({
      selectedIcon: obj ? obj.value : ''
    });
  };

  getSelectedIcon() {
    const { taskType } = this.props;

    return taskType ? taskType.icon : '';
  }

  renderOption = option => {
    return (
      <div className="icon-option">
        <Icon icon={option.value} />
        {option.label}
      </div>
    );
  };

  generateDoc = (values: { _id?: string; name: string }) => {
    const { taskType } = this.props;

    if (taskType) {
      values._id = taskType._id;
    }

    return {
      _id: values._id,
      name: values.name,
      icon: this.state.selectedIcon
    };
  };

  renderContent = (formProps: IFormProps) => {
    const { renderButton, closeModal, taskType } = this.props;
    const { values, isSubmitted } = formProps;
    const object = taskType || ({} as ITaskType);

    return (
      <>
        <FormGroup>
          <ControlLabel required={true}>Name</ControlLabel>
          <FormControl
            {...formProps}
            name="name"
            defaultValue={object.name}
            autoFocus={true}
            required={true}
          />
        </FormGroup>

        <FormGroup>
          <ControlLabel required={true}>Icon</ControlLabel>
          <Select
            isRequired={true}
            value={this.state.selectedIcon}
            options={TASK_ICONS}
            onChange={this.onChangeIcon}
            optionRenderer={this.renderOption}
            valueRenderer={this.renderOption}
          />
        </FormGroup>

        <Modal.Footer>
          <Button btnStyle="simple" onClick={closeModal} icon="cancel-1">
            Close
          </Button>

          {renderButton({
            name: 'task type',
            values: this.generateDoc(values),
            isSubmitted,
            callback: closeModal,
            object: taskType
          })}
        </Modal.Footer>
      </>
    );
  };

  render() {
    return <CommonForm renderContent={this.renderContent} />;
  }
}

export default Form;
