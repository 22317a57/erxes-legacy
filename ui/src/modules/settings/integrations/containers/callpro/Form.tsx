import ButtonMutate from 'modules/common/components/ButtonMutate';
import { IButtonMutateProps } from 'modules/common/types';
import CallPro from 'modules/settings/integrations/components/callpro/Form';
import { mutations } from 'modules/settings/integrations/graphql';
import React from 'react';

type Props = {
  type?: string;
  closeModal: () => void;
};

class CallProContainer extends React.Component<Props> {
  renderButton = ({
    name,
    values,
    isSubmitted,
    callback
  }: IButtonMutateProps) => {
    return (
      <ButtonMutate
        mutation={mutations.integrationsCreateExternalIntegration}
        variables={values}
        callback={callback}
        isSubmitted={isSubmitted}
        type="submit"
        successMessage={`You successfully added a ${name}`}
      />
    );
  };

  render() {
    const { closeModal } = this.props;

    const updatedProps = {
      closeModal,
      renderButton: this.renderButton
    };

    return <CallPro {...updatedProps} />;
  }
}

export default CallProContainer;
