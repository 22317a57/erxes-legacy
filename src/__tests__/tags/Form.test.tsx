import { mount, shallow } from 'enzyme';
import React from 'react';

import { IButtonMutateProps } from 'modules/common/types';
import Form from '../../modules/tags/components/Form';

describe('Form component', () => {
  const defaultProps = {
    type: 'typ',
    closeModal: () => null,
    save: (params: {
      doc: {
        _id?: string;
        name: string;
        type: string;
        colorCode: string;
      };
      callback: () => void;
    }) => null,
    afterSave: () => null,
    renderButton: (props: IButtonMutateProps) => <div />
  };

  test('renders Form successfully', () => {
    shallow(<Form {...defaultProps} />);
  });

  test('renders successfully with default value', () => {
    const wrapper = mount(<Form {...defaultProps} />);
    const props = wrapper.props();

    expect(props).toMatchObject(defaultProps);
  });
});
