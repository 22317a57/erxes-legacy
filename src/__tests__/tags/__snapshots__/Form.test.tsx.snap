// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component snapshot matches 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="sc-jKVCRD eDSGVr"
  >
    <label
      className="sc-ebFjAB bHuvQk"
    >
      Name
    </label>
    <input
      className="sc-kaNhvL bKDVoH"
      disabled={false}
      id="name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-jKVCRD eDSGVr"
  >
    <label
      className="sc-ebFjAB bHuvQk"
    >
      Color code
    </label>
    <input
      className="sc-kaNhvL bKDVoH"
      disabled={false}
      id="colorCode"
      onChange={[Function]}
      required={false}
      type="color"
      value="#8ff899"
    />
  </div>
  <div
    className="sc-jAaTju fBrEKz"
  >
    <button
      className="sc-ifAKCX iSmYrG"
      icon="cancel-1"
      type="button"
    >
      <i
        className="icon-cancel-1 sc-bwzfXH hpxbGw"
        icon="cancel-1"
      />
      <span>
        Cancel
      </span>
    </button>
    <button
      className="sc-ifAKCX cQdFDO"
      icon="checked-1"
      type="submit"
    >
      <i
        className="icon-checked-1 sc-bwzfXH hpxbGw"
        icon="checked-1"
      />
      <span>
        Save
      </span>
    </button>
  </div>
</form>
`;
